### ISSUE TITLE
Possible bug when passing varying object by reference to a function

### ISSUE THREAD

**Issue by Twinklebear** [ispc/ispc#2430]
The code below will fail to compile, with an error given about passing an lvalue to a `varying&` parameter:

```
struct Matrix {
    float m[16];
};

struct Object {
    Matrix m;
};

struct Instance {
    Object *o;
};

void transform(Matrix &m) {
}

void test(){
    Instance inst;
    transform(inst.o->m);
}
```

The error:
```
test.ispc:18:15: Error: Illegal to pass a "varying" lvalue to a reference parameter of type "varying struct Matrix &".
    transform(inst.o->m);
              ^^^^^^^^^
```

However it doesn't seem like I'm passing an lvalue here, the reference is taken to an object which exists in memory. What's also interesting is if I change the `o` member of `Instance` to be a value, the code compiles:

```
struct Matrix {
    float m[16];
};

struct Object {
    Matrix m;
};

struct Instance {
    Object o;
};

void transform(Matrix &m) {
}

void test(){
    Instance inst;
    transform(inst.o.m);
}
```

However nothing has changed about the varying/uniform-ness of the types or the indirection/memory references from what I can tell. So it seems like the error reported above when accessing it through a pointer is a bug.

**Comment by dbabokin** [comment]
The error message is misleading. Actually we do need an lvalue to pass it as a reference. This works the same way for uniform and varyings.

But for varying the value is an lvalue only if it is a continuous location in memory. It's possible (and it happens in this example) that what seems to be a varying lvalue, appears to be set of disjoint locations in memory and in this case it cannot be considered as an lvalue. This happens due to dereferencing a varying pointer.

In this example `struct Instance` is instantiated as varying and `Object *o` appears to be a varying pointer to uniform, i.e. `uniform Object * varying o`. So dereferencing it causes problem.

I think there's nothing to fix in issue in terms of compiler behavior, but the error message needs to be changed/improved.

**Comment by larsbohn** [comment]
Consider the following code:

https://godbolt.org/z/3MPrqcEhs

```
int f(int& num)
{
    return num;
}

void g()
{
    int s;              // Varying int
    int n[10];          // Array of varying ints
    int i;              // Varying int

    s += n[i];          // Ok
    s += f(i);          // Ok

    s += f(n[i]);       // Compile error
}
```

Is this a similar issue?

**Comment by aneshlya** [comment]
> Is this a similar issue?
 Yes, it is. It is important to have continuous location in memory.
https://godbolt.org/z/19KeT5Eeq
